### PR TITLE
Fix Flash Attention forward for Llama-2 70b

### LINKED ISSUE #28: question about the llama instruction code
Hello. The qlora and flash attention code has been very helpful.

I noticed that the code worked fine in llam2 7b, but there seems to be a shape issue in 70b. Is that correct?

### PR DESCRIPTION
**Pull request by davidmrau**
This pull request fixes #28  when training LLama-2 70b with LoRA and flash attention.

### CODE REVIEW

**Comment by davidmrau** [comment]
@philschmid can you have a look?

**Comment by philschmid** [comment]
Can you please explain why this should fix flash attention for GQA?

**Comment by philschmid** [comment]
The 70B are also having tp == 1 https://huggingface.co/meta-llama/Llama-2-70b-hf/blob/main/config.json#L17

File: training/utils/llama_patch.py [hunk 1]
```diff
@@ -38,16 +38,20 @@ def forward(
 
     attention_mask: [bsz, q_len]
     """
+    if self.config.pretraining_tp != 1:
+        raise ValueError("Only model.config.pretraining_tp=1 is supported but is", self.config.pretraining_tp)
```
Comment (philschmid) [inline training/utils/llama_patch.py:42]: wouldn't incude it here, slows down things and all pretrained checkpoints have the value as `1` on the hub.

**Review by philschmid — APPROVED** [review]
LGTM. Can you share a script to quickly reproduce it?

File: training/utils/llama_patch.py (hunk 1 above)
Comment (davidmrau) [reply training/utils/llama_patch.py]: sure, just removed it

**Comment by davidmrau** [comment]
> LGTM. Can you share a script to quickly reproduce it?

here you go: d91f6ddf2a0a05e92ebc865d8fe1e42b9eac7849

**Comment by davidmrau** [comment]
> > LGTM. Can you share a script to quickly reproduce it?
> 
> here you go: [d91f6dd](https://github.com/philschmid/deep-learning-pytorch-huggingface/commit/d91f6ddf2a0a05e92ebc865d8fe1e42b9eac7849)

actually, we don't want to have this in the repo so here: 

```python
from datasets import load_dataset
from peft import LoraConfig, prepare_model_for_kbit_training, get_peft_model
from transformers import TrainingArguments
from trl import SFTTrainer
from random import randrange
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

# Load dataset from the hub
dataset = load_dataset("databricks/databricks-dolly-15k", split="train")

print(f"dataset size: {len(dataset)}")
print(dataset[randrange(len(dataset))])
# dataset size: 15011


def format_instruction(sample):
	return f"""### Instruction:
Use the Input below to create an instruction, which could have been used to generate the input using an LLM.

### Input:
{sample['response']}

### Response:
{sample['instruction']}
"""


use_flash_attention = True
# COMMENT IN TO USE FLASH ATTENTION
# replace attention with flash attention
if torch.cuda.get_device_capability()[0] >= 8:
    from utils.llama_patch import replace_attn_with_flash_attn
    print("Using flash attention")
    replace_attn_with_flash_attn()


# Hugging Face model id
#model_id = "NousResearch/Llama-2-70b-hf" # non-gated
model_id = "meta-llama/Llama-2-70b-hf" # gated


# BitsAndBytesConfig int-4 config
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

# Load model and tokenizer
model = AutoModelForCausalLM.from_pretrained(model_id, quantization_config=bnb_config, use_cache=False, device_map="auto")
model.config.pretraining_tp = 1

# Validate that the model is using flash attention, by comparing doc strings
if use_flash_attention:
    from utils.llama_patch import forward
    assert model.model.layers[0].self_attn.forward.__doc__ == forward.__doc__, "Model is not using flash attention"


tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"




# LoRA config based on QLoRA paper
peft_config = LoraConfig(
        lora_alpha=16,
        lora_dropout=0.1,
        r=64,
        bias="none",
        task_type="CAUSAL_LM",
)

args = TrainingArguments(
    output_dir="llama-70-int4-dolly",
    num_train_epochs=3,
    per_device_train_batch_size=6 if use_flash_attention else 4,
    gradient_accumulation_steps=2,
    gradient_checkpointing=True,
    optim="paged_adamw_32bit",
    logging_steps=10,
    save_strategy="epoch",
    learning_rate=2e-4,
    bf16=True,
    tf32=True,
    max_grad_norm=0.3,
    warmup_ratio=0.03,
    lr_scheduler_type="constant",
#    disable_tqdm=True # disable tqdm since with packing values are in correct
)

# prepare model for training
model = prepare_model_for_kbit_training(model)

# Upcast layer for flash attnetion
if use_flash_attention:
    from utils.llama_patch import upcast_layer_for_flash_attention
    torch_dtype = torch.bfloat16 if args.bf16 else torch.float16 if args.fp16 else torch.float32
    model = upcast_layer_for_flash_attention(model, torch_dtype)


model = get_peft_model(model, peft_config)
max_seq_length = 2048 # max sequence length for model and packing of the dataset

trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    peft_config=peft_config,
    max_seq_length=max_seq_length,
    tokenizer=tokenizer,
    packing=True,
    formatting_func=format_instruction,
    args=args,
)


# train
trainer.train()
```

**Comment by philschmid** [comment]
Neeeded to revert this since the training for 7B became 10x worse and inference wasn't working anymore

**Comment by davidmrau** [comment]
For me inference with the 7B model works just the same

**Comment by davidmrau** [comment]
I will look into it next week when I have more time

**Comment by philschmid** [comment]
Yeah inference worked, was due flash attetnion being loaded. But running the example with the suggested patch from this branch results into poor performance. Will also try to see from what it is coming from if i have time.

**Pull request merged**